### PR TITLE
Code coverage improvements for alloc.h 

### PR DESCRIPTION
**Pull request by awelling2801**
**Work item:** Internal

**What were the changes?**  
Added Unit tests to increase code coverage for alloc.h
This PR adds one file - test/AllocTests.cpp comprising of 9 tests. 

**Why were the changes made?**  
The test improved code coverage for alloc.h 
|| Function Coverage  | Line Coverage | Region coverage | Branch Coverage |
|-------------| ------------- | ------------- | ------------- | ------------- |
| Default Coverage  |  57.89% (11/19)  |   67.19% (129/192)  |   56.78% (201/354) |   43.48% (60/138)  |
| Updated Coverage |  100.00% (19/19) |   94.79% (182/192) |   72.91% (261/358) |   52.90% (73/138)  |

**How was the outcome achieved?**  
The outcome was achieved by adding a AllocTests.cpp file comprising of the tests. 
The visibility set to default instead of hidden achieved by   https://github.com/ROCm/rccl/pull/1664
Also the code coverage was run on the Unit test binary and not librccl.so

	`/opt/rocm/lib/llvm/bin/llvm-cov show ./rccl-UnitTests --instr-profile=merged.profdata --format=html --output-dir=$WORKDIR/output/rccl_cov_report-lib --project-title=RCCL_Lib_Coverage_Report --object ../librccl.so`

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Comment by Nikhil-Nunna** [comment]
There is a failure in the Precheckin pipeline, looks like missing symbols

```2 warnings generated when compiling for host.
[100%] Linking CXX executable rccl-UnitTests
ld.lld: error: undefined symbol: ncclDebugNoWarn
>>> referenced by AllocTests.cpp
>>>               CMakeFiles/rccl-UnitTests.dir/AllocTests.cpp.o:(ncclResult_t ncclCudaMemcpy<float>(float*, float*, unsigned long))
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [test/CMakeFiles/rccl-UnitTests.dir/build.make:488: test/rccl-UnitTests] Error 1
make[1]: *** [CMakeFiles/Makefile2:153: test/CMakeFiles/rccl-UnitTests.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by venksubramd** [comment]
The test ncclIbMallocDebugZeroSize is good, it checks for the behavior correctly. The test ncclIbMallocDebugNonZero checks for the happy path. Do we need tests for the situations when the function ncclIbMalloc (that is ncclIbMallocDebug) would return ncclSystemError?

The test for ncclCuMemHostAlloc checks what happens if the input values are NULL (which is good), but we need test that checks what the actual behavior of the code is when all parameters are value. Also, for the negative test for NULL values, we should have multiple tests for different arguments being NULL instead of grouping all NULL arguments into one test.

Rest of the tests in this file, except the last one, check for error condition that results from invalid argument. We need tests that verify the expected behavior of these functions.

The file alloc.h has about two dozen functions, we should revisit and see which of those functions need test since right now we have tests only for a handfull of functions in alloc.h.

**Comment by awelling2801** [comment]
@venksubramd 
For tests concerning `ncclIbMalloc`, I could have a test where `ncclIbMalloc` returns `ncclSystemError`. This would require mocking `posix_memalign` which is tricky (not impossible) to do. Looking into this. 

`ncclCuMemHostAlloc` has two implementations covered by `#if CUDART_VERSION >= 12020`. For RCCL - any argument would return a `ncclInternalError` since it is just not supported. I do not think its worth it to explore this further unless this (and other functions which are still unsupported) have rccl support. 

> Rest of the tests in this file, except the last one, check for error condition that results from invalid argument. We need tests that verify the expected behavior of these functions

As mentioned, these are stubbed functions and the expected output for all of these functions is `ncclInternalError` 
 

> The file alloc.h has about two dozen functions, we should revisit and see which of those functions need test since right now we have tests only for a handfull of functions in alloc.h.

Function Coverage | Line Coverage | Region coverage | Branch Coverage
-- | -- | -- | --
Default Coverage | 57.89% (11/19) | 67.19% (129/192) | 56.78% (201/354) | 43.48% (60/138)
Updated Coverage | 100.00% (19/19) | 94.79% (182/192) | 72.91% (261/358) | 52.90% (73/138)

This coverage effectively validates the current implementation, and I’d consider the test a success.

**Comment by venksubramd** [comment]
@awelling2801 

Since the mocking efforts may take some time, it may be better to open a new ticket for that task. That way we can move forward to close this.

Thanks for your comments on parts that may not be relevant to rccl.

I am OK to close this if we open a new ticket for the mocking efforts.

File: test/AllocTests.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    }
+       
+    TEST(Alloc, ncclCuMemFreeAddr)
+    {
+        void* testPtr = reinterpret_cast<void*>(0xbeefcafe); // Arbitrary non-null pointer
+        ncclResult_t result = ncclCuMemFreeAddr(testPtr);
+        ASSERT_EQ(result, ncclInternalError);   
+    }
+    
+    TEST(Alloc, NcclCudaMemcpy) {
```
Comment (rahulvaidya20) [inline test/AllocTests.cpp:90]: It will be good if we can add more tests here to cover corner cases like `nullptr` handling, `0-byte copy`, etc.

**Review by rahulvaidya20 — COMMENTED** [review]
We need to revisit `cuMem` related tests once we have the support in ROCm 7.0.

File: test/AllocTests.cpp (hunk 1 above)
Comment (awelling2801) [reply test/AllocTests.cpp:90]: Added

**Comment by atulkulk** [comment]
> @venksubramd For tests concerning `ncclIbMalloc`, I could have a test where `ncclIbMalloc` returns `ncclSystemError`. This would require mocking `posix_memalign` which is tricky (not impossible) to do. Looking into this.
> 
> `ncclCuMemHostAlloc` has two implementations covered by `#if CUDART_VERSION >= 12020`. For RCCL - any argument would return a `ncclInternalError` since it is just not supported. I do not think its worth it to explore this further unless this (and other functions which are still unsupported) have rccl support.

@venksubramd  @awelling2801  Created a new ticket to look into adding new tests using GMock: [ GPUAI-5985 ]

**Review by rahulvaidya20 — APPROVED** [review]

**Review by atulkulk — APPROVED** [review]

**Comment by venksubramd** [comment]
@awelling2801 @atulkulk  Thanks for creating the new tickets. We can close this one.
